src/D2L.Security.OAuth2/Keys/KeyManagementService.cs

```diff
@@ -0,0 +1,207 @@
+
+			if( now > expectedNextRotation ) {
+				// If that's in the past, use a short retry window.
+				return TimeSpan.FromSeconds( 10 );
+			} else {
+				// Otherwise use that but with a little buffer for key
+				// generation time/imprecisely scheduled cron jobs.
+				return expectedNextRotation.AddMinutes( 1 ) - now;
+			}
+		}
+
+		async Task IKeyManagementService.GenerateNewKeyIfNeededAsync() {
+			var now = m_clock.UtcNow;
+
+			var keys = await m_privateKeys.GetAllAsync(
+				validUntilAtLeast: now
+			).ConfigureAwait( false );
+
+			foreach( var key in keys ) {
+				if( now < key.ExpiresAt - m_config.KeyRotationBuffer ) {
+					// This key is fine, do nothing.
+					return;
+				}
+			}
+
+			var keyId = Guid.NewGuid();
+			var keyIdStr = keyId.ToString();
+
+			using var csp = new RSACryptoServiceProvider {
+				PersistKeyInCsp = false
+			};
+
+			var pub = csp.ExportParameters(
+				includePrivateParameters: false
+			);
+
+			var priv = csp.ExportCspBlob(
+				includePrivateParameters: true
+			);
+
+			var expiresAt = now + m_config.KeyLifetime;
+
+			await Task.WhenAll(
+				m_publicKeys.SaveAsync(
+					keyId,
+					new RsaJsonWebKey(
+						keyIdStr,
+						expiresAt,
+						pub
+					)
+				),
+				m_privateKeys.SaveAsync(
+					new PrivateKeyData(
+						id: keyIdStr,
+						kind: PrivateKeyData.KeyKinds.Rsa,
+						data: priv,
+						createdAt: now,
+						notBefore: now + m_config.KeyTimeUntilUse,
+						expiresAt: expiresAt
+					)
+				)
+			).ConfigureAwait( false );
+		}
+
+		private async Task RefreshKeyAsync(
+			D2LSecurityToken current,
+			DateTimeOffset now
+		) {
+			var keys = await m_privateKeys.GetAllAsync(
+				validUntilAtLeast: now
+			).ConfigureAwait( continueOnCapturedContext: false );
+
+			var best = ChooseKey( keys, now ).Ref();
+
+			var loser = Interlocked.CompareExchange(
+				ref m_current,
+				best,
+				comparand: current
+			);
+
+			loser?.Dispose();
+		}
+
+		private D2LSecurityToken ChooseKey(
+			IEnumerable<PrivateKeyData> keys,
+			DateTimeOffset now
+		) {
+			PrivateKeyData candidate = null;
+
+			foreach( var key in keys ) {
+				// Ignore unsupported key types
+				if( key.Kind != PrivateKeyData.KeyKinds.Rsa ) {
+					continue;
+				}
+
+				// Use any non-expired key if it's the only one
+				if( candidate == null ) {
+					candidate = key;
+					continue;
+				}
+
+				if( candidate.NotBefore > now && key.NotBefore <= now ) {
+					// If we can switch to a key past its NotBefore, do that.
+					candidate = key;
+				} else if( candidate.NotBefore > now && candidate.CreatedAt > key.CreatedAt ) {
+					// When comparing two keys that are not past their respective
+					// NotBefore points, prefer the oldest one.
+					candidate = key;
+				} else if ( candidate.NotBefore <= now && key.NotBefore <= now && candidate.ExpiresAt < key.ExpiresAt ) {
+					// If we have two keys past their NotBefore date, pick the
+					// one that expires the latest.
+					candidate = key;
+				}

```

Comment:
Should this be created most recently?

Generate Key A, 3 month expiry
Adjust config, setting expiry to 1 week.
Generate Key B
Generate Key C
Generate Key D

B, C, D will all be ignored in favour of using A

Reply:
🤷‍♂️ I can see the case for that but I have no feelings (other than doing something reasonable and predictable.) I'll change it to that.